Assets/Package/Runtime/Tooltip/TooltipTarget.cs

```diff
@@ -10,29 +11,36 @@ namespace BroWar.UI.Tooltip
     /// </summary>
     [DisallowMultipleComponent]
     [AddComponentMenu("BroWar/UI/Tooltip/Tooltip Target")]
-    public class TooltipTarget : MonoBehaviour, IPointerEnterHandler, IPointerExitHandler
+    public abstract class TooltipTarget<T> : MonoBehaviour, IPointerEnterHandler, IPointerExitHandler where T : TooltipBehaviour
     {
-        [SerializeField, TextArea(4, 8)]
-        private string tooltipContent;
-        [SerializeField]
-        private TooltipSettings settings;
+        [SerializeField, FormerlySerializedAs("settings")]
+        private TooltipData data;
         [SerializeField, Min(0)]
         [Tooltip("Time needed to show tooltip content.")]
         private float offsetTime = 0.0f;
+        [SerializeField]
+        [Tooltip("Optional, custom tooltip prefab. If null, the default one will be used.")]
+        private T customPrefab;
 
         private ITooltipHandler tooltipHandler;
         private Sequence sequence;
 
-        private void ShowTooltip()
+        protected virtual void ShowTooltip()
         {
-            tooltipHandler.ShowTooltip(tooltipContent, in settings);
+            var tooltip = tooltipHandler.ShowTooltip(in data, customPrefab);
+            if (tooltip != null)
+            {
+                UpdateContent(tooltip);
```

Comment:
Niezbyt podoba mi się, że konfiguracja tooltipa jest rozbita na dwie klasy - najpierw handler, który za pomocą fabryki tworzy tooltip, i potem Target, który ustawia content. Też niezbyt mi pasuje, że pokazujemy obiekt, który nie jest przygotowany/skończony (metoda ShowTooltip pokazuje tooltipa). Wydaje mi się, że to sam TooltipBehaviour powinien zdecydować co zrobić z danymi, które dostaje, a TooltipTarget powinien być tylko kontenerem na dane. W takim wypadku to TooltipData dorzuciłbym content tooltipa.

Reply:
Nie zgadzam się. Początkowo tak zrobiłem ale jest tooltipy, czy tego chcemy czy nie, nie składają się po prostu z tekstu i nie zawsze składają się z tekstu statycznego. Ustawianie contentu to nie inicjalizacja, inicjalizacją jest ustawienie mu propert związanych z jego transformem i wiodkiem na Scenie. Content jest totalnie osobna kwestia bo możesz mieć tooltip, który ma tylko tekst, możesz mieć tooltip, który jest kompletnie statyczny, a możesz mieć też tooltip, który musi być aktualizowany co klatkę bo wyświetla np. o przychodach. Starałem się zrobić ten system możliwie elastyczny dlatego to Targety się implementuje by zarządzać contentem tooltipa i należy dziedziczyć odpowiednią metodą by to zrobić. StandardTooltipTarget jest tylko podstawową implementacją, która dodaje tekst i tyle.

Co do pokazywania obiektu, który nie jest przygotowany - nie rozumiem tego argumentu. Mógłbym dodać do TooltipHandlera metodę GetTooltip, potem zrobić UpdateContent i potem znowu musiałbym w TooltipHandlerze wykonać ShowTooltip i przekazać mu instancję tooltipa, która ma zostać pokazana. Nic to nie zmienia w sposobie działania systemu, a IMO komplikujemy bardzo proste API.

Reply:
No właśnie ale kiedy ustawiasz mu transform (np. żeby ustawić go w odpowiednim miejscu) to nie zrobisz tego zawsze dobrze nie znając rozmiaru recta. A ustawiasz go dopiero po pokazaniu i aktualizujesz w kolejnej klatce. Wiec jeżeli powiedzmy ustawisz sobie delay na 0 to pokażesz go w jakimś miejscu i na jedna klatkę będzie w złym miejscu bo nie miał jeszcze wtedy ustawionego tekstu. Z tego względu staram się zamykać wszystkie rzeczy z konfiguracja obiektu w jednym miejscu i wywoływać funkcje updateContent z zewnątrz w handlarze w tym wypadku, a nie bezpośrednio w targecie. I w tym wypadku zamiast TooltipData podawać TooltipTarget albo jakiś interfejs. Powiedz co myślisz.

Reply:
Pozycja nie jest ustawiana w następnej klatce, jest ustawiana w tej samej w metodzie UpdatePositionAndData. Generalnie to wolę mieć rozdzielone pozycjonowanie od ustawiania kontentu bo IMO to dwie osobne kwestie. To przemyślmy jakie mamy możliwości:

1. (obecna) Handler tworzy Tooltip i ustawia pozycję na bazie danych z Targetu, target dostaje zwrotkę, ustawia Content
2. Target pobiera sobie instancję Tooltipa, ustawia content i wywołuje Show w Handlerze z przygotowaną instancją
3. Target pobiera sobie instancję Tooltipa, ustawia content, ustawia pozycję (Pointer Position musi wziąć z Handlera) i wywołuje Show w Handlerze z przygotowaną instancją
4. Target pobiera sobie instancję Tooltipa, ustawia content oraz przekazuje interfejs odpowiedzialny za pozycjonowanie Tooltipa, instancja sama się ustawia i ma swój Update albo Update wywoływany z Handlera

Nie chciałbym żeby Target zarządzał dynamiczną pozycją bo będzie za dużo Update'ów. To która opcja Ci najbardziej odpowiada i czy masz inną?


 
- 
-

Reply:
Dogadaliśmy już